src/models/transfers.js

```diff
@@ -207,31 +208,83 @@ function * processStateTransitions (tr, transfer) {
       yield accountBalances.applyDebits()
       updateState(transfer, 'prepared')
     }
+
+    // If the transfer is prepared and has no execution condition, execute immediately
+    if (transfer.state === 'prepared' && !Boolean(transfer.execution_condition)) {
+      yield accountBalances.applyCredits()
+
+      // Remove the expiry countdown
+      transferExpiryMonitor.unwatch(transfer.id)
+      updateState(transfer, 'executed')
+    }
   }
+}
 
-  if (transfer.state === 'prepared' && transfer.hasFulfillment('execution')) {
-    if (!transfer.hasValidFulfillment('execution')) {
-      throw new UnmetConditionError('ConditionFulfillment failed')
+function * fulfillTransfer (transferId, fulfillment) {
+  let transfer
+  yield db.transaction(function *(transaction) {
+    transfer = yield db.getTransfer(transferId, {transaction})
+
+    if (!transfer) {
+      throw new NotFoundError('Invalid transfer ID')
     }
-    yield accountBalances.applyCredits()
 
-    // Remove the expiry countdown
-    transferExpiryMonitor.unwatch(transfer.id)
+    let accountBalances = yield makeAccountBalances(transaction, transfer)
 
-    updateState(transfer, 'executed')
-  }
+    // We don't know if notary is executing or rejecting transfer
+    const isValidExecution = Boolean(transfer.execution_condition && Condition.testFulfillment(transfer.execution_condition, fulfillment))
+    const isValidCancellation = Boolean(transfer.cancellation_condition && Condition.testFulfillment(transfer.cancellation_condition, fulfillment))
 
-  let canRejectState = transfer.state === 'proposed' || transfer.state === 'prepared'
-  if (canRejectState && transfer.cancellation_condition_fulfillment) {
-    if (!transfer.hasValidFulfillment('cancellation')) {
+    // Exactly one condition must be true, so booleans must sum to 1
+    if (isValidCancellation + isValidExecution !== 1) {
```

Comment:
`isValidCancellation === isValidExecution` is simpler and I think clear enough to remove the need for comment